base-files: add bootup-time measurement

feckert opened the pull request:
In order to measure the boot time, it must be set during the boot process with the last init script.
The last boot script that comes into question from my point of view is the `done` init script.

This pull request moves all start scripts that are already at 99 forward by one and moves the done script from position 95 to 99, as it must run last.

The done script was extended so that it writes the uptime to a file.
The saved value shows how long the boot process took.

Ansuel commented on package/base-files/files/etc/init.d/done:
```diff
@@ -11,6 +11,13 @@ boot() {
 		sh /etc/rc.local
 	}
 
+	# saveing bootup-time
```
saving? but from what i can read here we read it?

feckert commented:
@Ansuel Apart from the other feeds (where there is still START=99), I have not found anything else in the main repository that is on 99.

> Would be so good to unify all of that (but wonder if that can cause some problems)

What about `/tmp/sysinfo`? We already have a few files in this directory.
The values could then also be read out easily with the `ubus call system info` or `ubus call system board`.

jow- commented:
Wouldn't [procd's rcS.c](https://lxr.openwrt.org/source/procd/rcS.c#L109) the far better place to implement such functionality? It would also be independent from any kind of init script pecularities. It appears there even already is timing code, just guarded by debug macros.

The measured value could then be exposed in the system/info ubus procedure.

Ansuel commented:
@jow- nice idea something external than script would also reduce the delta of this pr as we don't need to edit any script at all

feckert commented:
@jow- I have looked at your suggestion again.
That we process the whole thing in the `procd` and not in a `initscript` is a very good suggestion.
But the intention behind my PR is to measure the **whole** boot time, and not the boot time of the individual initscripts.
The kernel is also booted and loaded before `procd`, and the time also belongs to the boot time according to my understanding.
By my definition, this is from power on until all services are started.

If I understand correctly, this is the case when the procd has reached the state running ([STATE_RUNNING](https://lxr.openwrt.org/source/procd/state.c#L171)).

Therefore, I would suggest that we save the `uptime` at this state and make this available via ubus.

feckert committed:
procd: add bootup time measurement

Until now it was not possible to determine the booting time. This change
adds two new elements to the 'ubus system info'.

booting:
This flag indicates whether the system is still booting.

bootime:
This value shows in seconds how long the system took until the procd
reached STATE_INIT. The procd message in the log with '- init -'
indicates this.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert commented:
@Ansuel @jow- I have now implemented this as I imagine it in procd to measure the booting time from power-on to procd init complete.

ynezz commented:
> I have now implemented this as I imagine it in procd

BTW this can't be merged as it is, you need to send it as a patch "[PATCH procd]" to [the mailing list](https://openwrt.org/docs/guide-developer/working-with-git-email).

ynezz commented on package/system/procd/patches/0001-ubus-add-bootup-time-measurement.patch:
```diff
@@ -0,0 +1,124 @@
+Subject: [PATCH] ubus: add bootup time measurement
+
+Until now it was not possible to determine the booting time. This change
+adds two new elements to the 'ubus system info'.
```
I would probably think more about other possible use cases and would try to make it more flexible, instead of just focusing on one specific use case, probably something like the following:

```json
{
      ...
      "init_state": {
          "early" : { "duration": 23.058, "active": false },
          "ubus" : { "duration": 3.52, "active": false },
          "init" : { "duration": 15.22, "active": false },
          "running" : { "duration": 3457.89, "active": true },
          "shutdown": { "duration": 0, "active": false },
      }
}
```
> This flag indicates whether the system is still booting.

So in your use case, you would know, that the system is booting if `running.active` is `false`.

ynezz commented on package/system/procd/patches/0001-ubus-add-bootup-time-measurement.patch:
```diff
@@ -0,0 +1,124 @@
+This flag indicates whether the system is still booting.
+
+bootime:
+This value shows in seconds how long the system took until the procd
```
> This value shows in seconds

It's too coarse, increase of 1s is a lot of time during boot process, I would go with milliseconds.

feckert commented:
@ynezz friendly ping for https://patchwork.ozlabs.org/project/openwrt/patch/20221026131851.644176-1-fe@dev.tdt.de/
